File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.cpp
class Solution {
public:
    
    int foo(vector<int>& cost, int index, vector<int>& cache){
        
        int n=cost.size()-1;
        
        if(cache[index]!=-1){
            return cache[index];
        }
        
        //base case
        if(index==n || index==n-1){
            return cost[index];
        }
        
        long int cst1=foo(cost,index+1,cache);
        long int cst2=foo(cost,index+2,cache);
        cache[index+2]=cst2;
        cache[index+1]=cst1;
        
        cache[index]=cost[index]+min(cst1,cst2);
        
        return cache[index];
        
    }
    
    
    int minCostClimbingStairs(vector<int>& cost) {
            vector<int> cache(cost.size(),-1);
            return min(foo(cost,0,cache),foo(cost,1,cache));
    }
};

File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        n=len(cost)
        if(n<=1):
            return min(cost[0],cost[1])
        
        def foo(idx):
            if(idx==n-1 or idx==n-2):
                return cost[idx]
            return cost[idx]+min(foo(idx+1),foo(idx+2))
        
        def fooDP():
            dp=[0]*(n+1)
            
            dp[0]=cost[0]
            dp[1]=cost[1]
            
            for i in range(2,n):
                dp[i]=cost[i]+min(dp[i-1],dp[i-2])
            
            return min(dp[n-2],dp[n-1])
        
        
        return fooDP()
    
        
    
        